### ISSUE TITLE
Support hash parameter in v-link and improve custom handling

### ISSUE THREAD

**Issue by noirbizarre** [vuejs/vue-router#434]
It would be nice to be able to use `hash` with named route pattern.

### Currently
In vue-router 0.7.11, in History/html5 mode:
- hash is automatically extracted from path given as string and scroll to this hash after route change
  (ie. `v-link="/page#section"` will to to the route `/page` and scroll to the element with this id (https://github.com/vuejs/vue-router/blob/master/src/index.js#L543-L560)
- there is no way to specify an anchor/a hash with the named route pattern (ie. `v-link="{name: 'page', params: {}}"`)

### Proposal
- Add/handle an optional `hash` or `anchor` in `v-link` named route pattern handling (ie. `v-link="{name: 'page', hash: 'section'}"`)
- Add a, optional `hash()` transition hook called after `data` hook, if hash value change (same behavior as `data` hook for hash) :
```javascript
Vue.component('hash-hook-example', {
  // ... other options
  route: {
    hash (value) {
      console.log(`hash-hook triggered with ${value}`)
    }
  }
})
```
If the hook is present, it is called. If not, the default behavior (aka. scroll to `getElementById(hash)`) is executed

**Comment by divmgl** [comment]
+1 on this, would really like to see links that are create by Vue and can scroll to an anchor

**Comment by fnlctrl** [comment]
@noirbizarre @divmgl Scrolling to hash is implemented in 2.0, as in [scrollBehavior](https://github.com/vuejs/vue-router/blob/43183911dedfbb30ebacccf2d76ced74d998448a/examples/scroll-behavior/app.js#L18-L38) ([release notes](https://github.com/vuejs/vue-router/releases/tag/v2.0.0-beta.1)),

As for hash hook, since all transition hooks are deprecated in 2.0, there won't be any new hooks added, so you can just use a watcher on $route.
